File: src/Core/PromptManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;
using PromptSpec.Models;
using PromptSpec.Exceptions;

namespace PromptSpec.Core
{

    /// <summary>
    /// Manages loading and retrieving prompt templates from YAML configuration.
    /// </summary>
    public class PromptManager
    {
    private readonly Dictionary<string, PromptTemplate> _templates = new Dictionary<string, PromptTemplate>();
        private readonly IDeserializer _deserializer;

        /// <summary>
        /// Initializes a new instance of the PromptManager class.
        /// </summary>
        public PromptManager()
        {
            _deserializer = new DeserializerBuilder()
                .WithNamingConvention(CamelCaseNamingConvention.Instance)
                .IgnoreUnmatchedProperties()
                .Build();
        }

        /// <summary>
        /// Gets the count of loaded templates.
        /// </summary>
        public int TemplateCount => _templates.Count;

        /// <summary>
        /// Gets the names of all loaded templates.
        /// </summary>
        public IEnumerable<string> TemplateNames => _templates.Keys;

        /// <summary>
        /// Parses the YAML content and loads all prompts into memory.
        /// </summary>
        /// <param name="yamlContent">The YAML content containing prompt definitions.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        /// <exception cref="ArgumentNullException">Thrown when yamlContent is null.</exception>
        /// <exception cref="PromptValidationException">Thrown when YAML parsing or validation fails.</exception>
        public async Task LoadTemplatesAsync(string yamlContent)
        {
            if (string.IsNullOrEmpty(yamlContent))
            {
                throw new ArgumentException("YAML content cannot be null or empty.", nameof(yamlContent));
            }

            try
            {
                await Task.Run(() =>
                {
                    var promptCollection = _deserializer.Deserialize<PromptCollection>(yamlContent);

                    // If deserialization returns null (e.g., whitespace-only content), throw exception
                    if (promptCollection == null)
                    {
                        throw new PromptValidationException("Invalid YAML: No 'prompts' section found or it is null.");
                    }

                    // If prompts section is missing entirely, just clear templates (success case)
                    // If prompts is explicitly set to null, throw exception
                    if (promptCollection.Prompts == null)
                    {
                        // Check if the YAML explicitly contains "prompts:" (even if null)
                        // If it does, throw exception. If it doesn't, succeed.
                        if (yamlContent.Contains("prompts:"))
                        {
                            throw new PromptValidationException("Invalid YAML: No 'prompts' section found or it is null.");
                        }
                        else
                        {
                            _templates.Clear();
                            return;
                        }
                    }

                    _templates.Clear();

                    foreach (var specification in promptCollection.Prompts)
                    {
                        ValidateSpecification(specification);

                        if (_templates.ContainsKey(specification.Name))
                        {
                            throw new PromptValidationException($"Duplicate prompt name '{specification.Name}' found.");
                        }

                        _templates[specification.Name] = new PromptTemplate(specification);
                    }
                });
            }
            catch (YamlDotNet.Core.YamlException ex)
            {
                throw new PromptValidationException("Failed to parse YAML content.", (Exception)ex);
            }
            catch (Exception ex)
            {
                if (!(ex is PromptValidationException))
                {
                    throw new PromptValidationException("An error occurred while loading templates.", ex);
                }
                throw;
            }
        }

        /// <summary>
        /// Loads templates from a YAML file.
        /// </summary>
        /// <param name="filePath">Path to the YAML file.</param>
        /// <returns>A task representing the asynchronous operation.</returns>
        /// <exception cref="FileNotFoundException">Thrown when the specified file is not found.</exception>
        /// <exception cref="PromptValidationException">Thrown when YAML parsing or validation fails.</exception>
        public async Task LoadTemplatesFromFileAsync(string filePath)
        {
            if (!File.Exists(filePath))
            {
                throw new FileNotFoundException($"Template file not found: {filePath}");
            }

            string yamlContent;
            using (var reader = new System.IO.StreamReader(filePath))
            {
                yamlContent = reader.ReadToEnd();
            }
            await LoadTemplatesAsync(yamlContent);
        }

        /// <summary>
        /// Retrieves a specific prompt template by its name.
        /// </summary>
        /// <param name="promptName">The name of the prompt to retrieve.</param>
        /// <returns>The prompt template, or null if not found.</returns>
        /// <exception cref="ArgumentNullException">Thrown when promptName is null.</exception>
    public PromptTemplate GetPrompt(string promptName)
        {
            if (string.IsNullOrEmpty(promptName))
            {
                throw new ArgumentException("Prompt name cannot be null or empty.", nameof(promptName));
            }
            PromptTemplate template;
            return _templates.TryGetValue(promptName, out template) ? template : null;
        }

        /// <summary>
        /// Retrieves a specific prompt template by its name and throws an exception if not found.
        /// </summary>
        /// <param name="promptName">The name of the prompt to retrieve.</param>
        /// <returns>The prompt template.</returns>
        /// <exception cref="ArgumentNullException">Thrown when promptName is null.</exception>
        /// <exception cref="PromptNotFoundException">Thrown when the prompt is not found.</exception>
        public PromptTemplate GetRequiredPrompt(string promptName)
        {
            var template = GetPrompt(promptName);
            if (template == null)
            {
                throw new PromptNotFoundException(promptName);
            }
            return template;
        }

        /// <summary>
        /// Checks if a prompt template with the specified name exists.
        /// </summary>
        /// <param name="promptName">The name of the prompt to check.</param>
        /// <returns>True if the prompt exists, false otherwise.</returns>
        public bool HasPrompt(string promptName)
        {
            return !string.IsNullOrEmpty(promptName) && _templates.ContainsKey(promptName);
        }

        /// <summary>
        /// Clears all loaded templates.
        /// </summary>
        public void Clear()
        {
            _templates.Clear();
        }

        /// <summary>
        /// Gets all loaded prompt templates.
        /// </summary>
        /// <returns>Collection of all prompt templates.</returns>
        public IEnumerable<PromptTemplate> GetAllPrompts()
        {
            return _templates.Values;
        }

        /// <summary>
        /// Validates a prompt specification for required fields and consistency.
        /// </summary>
        /// <param name="specification">The specification to validate.</param>
        /// <exception cref="PromptValidationException">Thrown when validation fails.</exception>
        private static void ValidateSpecification(PromptSpecification specification)
        {
            if (string.IsNullOrEmpty(specification.Name))
            {
                throw new PromptValidationException("Prompt name is required and cannot be empty.");
            }

            if (string.IsNullOrEmpty(specification.Template))
            {
                throw new PromptValidationException($"Template is required for prompt '{specification.Name}' and cannot be empty.");
            }

            // Validate parameter ranges
            if (specification.Parameters != null)
            {
                if (specification.Parameters.Temperature.HasValue)
                {
                    var temp = specification.Parameters.Temperature.Value;
                    if (temp < 0 || temp > 2.0)
                    {
                        throw new PromptValidationException($"Temperature for prompt '{specification.Name}' must be between 0 and 2.0.");
                    }
                }

                if (specification.Parameters.TopP.HasValue)
                {
                    var topP = specification.Parameters.TopP.Value;
                    if (topP < 0 || topP > 1.0)
                    {
                        throw new PromptValidationException($"TopP for prompt '{specification.Name}' must be between 0 and 1.0.");
                    }
                }

                if (specification.Parameters.MaxTokens.HasValue && specification.Parameters.MaxTokens.Value <= 0)
                {
                    throw new PromptValidationException($"MaxTokens for prompt '{specification.Name}' must be greater than 0.");
                }
            }

            // Validate placeholder definitions
            if (specification.Placeholders != null)
            {
                foreach (KeyValuePair<string, PlaceholderDefinition> kvp in specification.Placeholders)
                {
                    string placeholderName = kvp.Key;
                    PlaceholderDefinition definition = kvp.Value;
                    if (string.IsNullOrEmpty(placeholderName))
                    {
                        throw new PromptValidationException($"Placeholder name cannot be empty in prompt '{specification.Name}'.");
                    }
                    if (!IsValidPlaceholderType(definition.Type))
                    {
                        throw new PromptValidationException($"Invalid placeholder type '{definition.Type}' for placeholder '{placeholderName}' in prompt '{specification.Name}'. Valid types are: string, number, boolean.");
                    }
                }
            }
        }

        /// <summary>
        /// Validates if a placeholder type is supported.
        /// </summary>
        /// <param name="type">The type to validate.</param>
        /// <returns>True if the type is valid, false otherwise.</returns>
        private static bool IsValidPlaceholderType(string type)
        {
            if (type == null)
                return false;
            var t = type.ToLowerInvariant();
            return t == "string" || t == "number" || t == "boolean";
        }
    }
}

File: python/promptspec/core/prompt_manager.py
"""
Manages loading and retrieving prompt templates from YAML configuration.
"""
import yaml
from pathlib import Path
from typing import Dict, Optional, List, Union
from ..models.prompt_collection import PromptCollection
from ..models.prompt_specification import PromptSpecification
from .prompt_template import PromptTemplate
from ..exceptions.prompt_exceptions import (
    PromptValidationException, 
    PromptNotFoundException
)


class PromptManager:
    """Manages loading and retrieving prompt templates from YAML configuration."""
    
    def __init__(self):
        self._templates: Dict[str, PromptTemplate] = {}
    
    @property
    def template_count(self) -> int:
        """Gets the count of loaded templates."""
        return len(self._templates)
    
    @property
    def template_names(self) -> List[str]:
        """Gets the names of all loaded templates."""
        return list(self._templates.keys())
    
    async def load_templates_async(self, yaml_content: str) -> None:
        """
        Parses YAML content and loads all prompts into memory.
        
        Args:
            yaml_content: The YAML content containing prompt definitions
            
        Raises:
            ValueError: When yaml_content is None or empty
            PromptValidationException: When YAML parsing or validation fails
        """
        if not yaml_content or not yaml_content.strip():
            raise ValueError("yaml_content cannot be None or empty")
        
        try:
            # Parse YAML content
            yaml_data = yaml.safe_load(yaml_content)
            
            if not yaml_data or 'prompts' not in yaml_data:
                raise PromptValidationException(
                    "Invalid YAML: No 'prompts' section found or it is null."
                )
            
            # Validate and create prompt collection using Pydantic
            prompt_collection = PromptCollection.model_validate(yaml_data)
            
            self._templates.clear()
            
            for specification in prompt_collection.prompts:
                self._validate_specification(specification)
                
                if specification.name in self._templates:
                    raise PromptValidationException(
                        f"Duplicate prompt name '{specification.name}' found."
                    )
                
                self._templates[specification.name] = PromptTemplate(specification)
                
        except yaml.YAMLError as e:
            raise PromptValidationException("Failed to parse YAML content.") from e
        except Exception as e:
            if isinstance(e, PromptValidationException):
                raise
            raise PromptValidationException("An error occurred while loading templates.") from e
    
    def load_templates(self, yaml_content: str) -> None:
        """Synchronous version of template loading."""
        if not yaml_content or not yaml_content.strip():
            raise ValueError("yaml_content cannot be None or empty")
        
        try:
            # Parse YAML content
            yaml_data = yaml.safe_load(yaml_content)
            
            if not yaml_data or 'prompts' not in yaml_data:
                raise PromptValidationException(
                    "Invalid YAML: No 'prompts' section found or it is null."
                )
            
            # Validate and create prompt collection using Pydantic
            prompt_collection = PromptCollection.model_validate(yaml_data)
            
            self._templates.clear()
            
            for specification in prompt_collection.prompts:
                self._validate_specification(specification)
                
                if specification.name in self._templates:
                    raise PromptValidationException(
                        f"Duplicate prompt name '{specification.name}' found."
                    )
                
                self._templates[specification.name] = PromptTemplate(specification)
                
        except yaml.YAMLError as e:
            raise PromptValidationException("Failed to parse YAML content.") from e
        except Exception as e:
            if isinstance(e, PromptValidationException):
                raise
            raise PromptValidationException("An error occurred while loading templates.") from e
    
    async def load_templates_from_file_async(self, file_path: Union[str, Path]) -> None:
        """
        Loads templates from a YAML file.
        
        Args:
            file_path: Path to the YAML file
            
        Raises:
            FileNotFoundError: When the specified file is not found
            PromptValidationException: When YAML parsing or validation fails
        """
        path = Path(file_path)
        
        if not path.exists():
            raise FileNotFoundError(f"Template file not found: {file_path}")
        
        yaml_content = path.read_text(encoding='utf-8')
        await self.load_templates_async(yaml_content)
    
    def load_templates_from_file(self, file_path: Union[str, Path]) -> None:
        """Synchronous version of template loading from file."""
        path = Path(file_path)
        
        if not path.exists():
            raise FileNotFoundError(f"Template file not found: {file_path}")
        
        yaml_content = path.read_text(encoding='utf-8')
        self.load_templates(yaml_content)
    
    def get_prompt(self, prompt_name: str) -> Optional[PromptTemplate]:
        """
        Retrieves a specific prompt template by its name.
        
        Args:
            prompt_name: The name of the prompt to retrieve
            
        Returns:
            The prompt template, or None if not found
            
        Raises:
            ValueError: When prompt_name is None or empty
        """
        if not prompt_name or not prompt_name.strip():
            raise ValueError("prompt_name cannot be None or empty")
        
        return self._templates.get(prompt_name)
    
    def get_required_prompt(self, prompt_name: str) -> PromptTemplate:
        """
        Retrieves a specific prompt template by name, raising exception if not found.
        
        Args:
            prompt_name: The name of the prompt to retrieve
            
        Returns:
            The prompt template
            
        Raises:
            ValueError: When prompt_name is None or empty
            PromptNotFoundException: When the prompt is not found
        """
        template = self.get_prompt(prompt_name)
        if template is None:
            raise PromptNotFoundException(prompt_name)
        return template
    
    def has_prompt(self, prompt_name: str) -> bool:
        """
        Checks if a prompt template with the specified name exists.
        
        Args:
            prompt_name: The name of the prompt to check
            
        Returns:
            True if the prompt exists, false otherwise
        """
        return bool(prompt_name and prompt_name.strip() and prompt_name in self._templates)
    
    def clear(self) -> None:
        """Clears all loaded templates."""
        self._templates.clear()
    
    def get_all_prompts(self) -> List[PromptTemplate]:
        """Gets all loaded prompt templates."""
        return list(self._templates.values())
    
    def _validate_specification(self, specification: PromptSpecification) -> None:
        """
        Validates a prompt specification for required fields and consistency.
        
        Args:
            specification: The specification to validate
            
        Raises:
            PromptValidationException: When validation fails
        """
        # Pydantic already validates basic structure, but add custom business logic
        if not specification.name.strip():
            raise PromptValidationException("Prompt name is required and cannot be empty.")
        
        if not specification.template.strip():
            raise PromptValidationException(
                f"Template is required for prompt '{specification.name}' and cannot be empty."
            )
        
        # Additional validation for placeholder definitions
        if specification.placeholders:
            for placeholder_name, definition in specification.placeholders.items():
                if not placeholder_name.strip():
                    raise PromptValidationException(
                        f"Placeholder name cannot be empty in prompt '{specification.name}'."
                    )